File: analyzers/tests/SonarAnalyzer.Test/TestCases/ExceptionsShouldBePublic.vb
Imports System.Collections.Generic

Public Class PublicException ' Compliant
    Inherits Exception
End Class

Friend Class InternalException
    '        ^^^^^^^^^^^^^^^^^ Noncompliant {{Make this exception 'Public'.}}
    Inherits Exception
End Class

Class ImplicitAccessibilityException
    ' ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Noncompliant {{Make this exception 'Public'.}}
    Inherits Exception

End Class

Public Class PublicContainer
    Public Class PublicException ' Compliant
        Inherits Exception
    End Class

    Private Class PrivateClass ' Compliant, no exception
    End Class

    Friend Class InternalException ' Noncompliant
        Inherits Exception
    End Class

    Class ImplicitAccessibilityException ' Compliant
        Inherits Exception
    End Class

    Protected Class InheritsFromOtherException ' Compliant
        Inherits DivideByZeroException
    End Class

    Private Class PrivateException ' Noncompliant
        Inherits Exception
    End Class
End Class

Friend Class InternalContainer
    Public Class PublicException
        '        ^^^^^^^^^^^^^^^ Noncompliant {{Make this exception 'Public'.}}
        Inherits Exception
    End Class

    Private Class PrivateClassn ' Compliant
    End Class

    Protected Class ProtectedClass ' Compliant
    End Class

    Friend Class InternalException
        Inherits DivideByZeroException ' Compliant
    End Class

    Protected Class ProtectedException ' Compliant
        Inherits DivideByZeroException
    End Class

    Class ImplicitAccessibilityException ' Noncompliant
        Inherits ApplicationException
    End Class

    Private Class PrivateException ' Noncompliant
        Inherits Exception
    End Class
End Class

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ExceptionsShouldBePublic.cs
using System;
using System.Collections.Generic;

public class PublicException : Exception // Compliant
{
}

internal class InternalException : Exception
//             ^^^^^^^^^^^^^^^^^ Noncompliant {{Make this exception 'public'.}}
{
}

class InternalException2 : Exception
//    ^^^^^^^^^^^^^^^^^^ Noncompliant {{Make this exception 'public'.}}
{
}

public class PublicContainer // Compliant
{
    public class PublicException : Exception // Compliant
    {
    }

    private class PrivateClass // Compliant
    {
    }

    internal class InternalException : Exception
//                 ^^^^^^^^^^^^^^^^^ Noncompliant {{Make this exception 'public'.}}
    {
    }

    class InternalException2 : Exception
//        ^^^^^^^^^^^^^^^^^^ Noncompliant {{Make this exception 'public'.}}
    {
    }

    protected class ProtectedException : OutOfMemoryException // Compliant
    {
    }

    private class PrivateException2 : Exception
//                ^^^^^^^^^^^^^^^^^ Noncompliant {{Make this exception 'public'.}}
    {
    }
}

internal class InternalContainer // Compliant
{
    public class PublicException : Exception
//               ^^^^^^^^^^^^^^^ Noncompliant {{Make this exception 'public'.}}
    {
    }

    private class PrivateClass // Compliant
    {
    }

    protected class ProtectedClass // Compliant
    {
    }

    internal class InternalException : OutOfMemoryException // Compliant
    {
    }

    protected class ProtectedException : OutOfMemoryException // Compliant
    {
    }

    class InternalException2 : ApplicationException
//        ^^^^^^^^^^^^^^^^^^ Noncompliant {{Make this exception 'public'.}}
    {
    }

    private class PrivateException2 : Exception
//                ^^^^^^^^^^^^^^^^^ Noncompliant {{Make this exception 'public'.}}
    {
    }
}

internal class LastException : MiddleException { }
public class MiddleException : FirstException { }
public class FirstException : Exception { }

namespace ShouldNotThrow
{
    public class /* Missing identifier */ : Exception { } // Error [CS1001]
}
